Adding chunk & type information to dask high level graphs

GenevieveBuckley opened the pull request:
This PR is to give us a place for discussing implementation strategies for https://github.com/dask/dask/issues/7141#issuecomment-784562777

@jrbourbeau - Matt says you'd probably like to have this discussion with me, or will tag in somebody else to handle it.
cc @madsbk, another person who might also be interested.

Steps:
1. Add info about the `chunks` & `dtype` to the dask high level graph, for both `Array` and `Dataframe`
2. Work out which tests are now failing & fix them (this is where we're at right now)
3. Think about how we could use this information for better high level graph visualizations (probably in a separate PR)

GenevieveBuckley commented:
Copying over info from https://github.com/dask/dask/issues/7141#issuecomment-787720414

> Brief update:
> I've added three lines to to dask Array `__new__` (I think this is what Matt was suggesting above)
> 
> ```python
> class Array(DaskMethodsMixin):
>     ...
>     def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
>         ...
>         self.dask.layers[name].type = type(self)
>         self.dask.layers[name].chunks = chunks
>         self.dask.layers[name].dtype = dtype
>         return self
> ```
> 
> Test failures look localized in two different areas.
> ### Failing tests from `test_array_core.py`
> 
> Running pytest, there are 7 failures in `pytest dask/array/tests/test_array_core.py -v`
> 
> The failures are almost all happening because `name` isn't a valid key in `self.dask.layers` for those particular tests. Usually, that's because we have `name = 'some-string'` and the key is a tuple like `('some-string', 0, 1)`.
> 
> In one test (`test_constructor_plugin`), it says that `self` is actually a numpy array instead and gives us `AttributeError: 'numpy.ndarray' object has no attribute 'dask'`.
> ### Failing tests from `test_distributed.py`
> 
> There are 14 failing tests if you run pytest on just this subset: `pytest dask/tests/test_distributed.py -v`
> 
> Oddly, if you run the whole pytest suite then you get 10 ERRORS from `test_distributed.py` but no actual test failures reported (maybe they're all skipped or something, idk).
> 
> I haven't looked at what's going wrong with these yet.

GenevieveBuckley commented:
This has been slower going than I'd like, because I keep getting failing tests locally when I run pytest on the master branch. Some of it is this https://github.com/dask/dask/issues/7291 but my workaround of increasing `ulimit -n` hasn't quite resolved everything.

GenevieveBuckley commented:
> This has been slower going than I'd like, because I keep getting failing tests locally when I run pytest on the master branch. Some of it is this #7291 but my workaround of increasing `ulimit -n` hasn't quite resolved everything.

Update: deleting everything (github repo & conda env) then starting again did the trick.

GenevieveBuckley commented:
So I think for dataframes the best place is to edit the `new_dd_object` function.

https://github.com/dask/dask/blob/8663c6b7813fbdcaaa85d4fdde04ff42b1bb6ed0/dask/dataframe/core.py#L6485

Are dataframe divisions analogous to dask array chunks?

I think I can find the dtype by looking at what `meta` is.

GenevieveBuckley commented on dask/dataframe/core.py:
```diff
@@ -3609,6 +3609,12 @@ class DataFrame(_Frame):
+    def __init__(self, dsk, name, meta, divisions):
+        super().__init__(dsk, name, meta, divisions)
+        self.dask.layers[name].type = type(self)
+        self.dask.layers[name].divisions = divisions
```
I think `divisions` is the dataframe analogue to `chunks` in dask arrays - please correct me if I'm wrong here.

GenevieveBuckley commented on dask/dataframe/core.py:
```diff
@@ -3609,6 +3609,12 @@ class DataFrame(_Frame):
+        super().__init__(dsk, name, meta, divisions)
+        self.dask.layers[name].type = type(self)
+        self.dask.layers[name].divisions = divisions
+        self.dask.layers[name].chunk_type = type(meta)
```
There's no `dtype` attribute for dataframes.

Do we want to include information about the type of what's contained in the Dask array? (If the answer here is yes, then I can add something similar for arrays in this PR)

GenevieveBuckley commented:
The tests all pass if I run pytest with only the changes made to `dask/dataframe/core.py`. So that's an unexpected bonus, I was expecting to find more problems pop up when I did that.

GenevieveBuckley commented:
Also, the errors/failures I saw earlier in `test_distributed.py` seem to have disappeared (maybe related to whatever weirdness was causing [this](https://github.com/dask/dask/pull/7309#issuecomment-789584275)). So that's good news too.

GenevieveBuckley commented:
Found the common thread!
These changes aren't compatible with `dask.array.store()` given a `compute=False` keyword argument.

GenevieveBuckley commented:
Here is a better minimal reproducible example for our test failures (example is [from the docs here](https://docs.dask.org/en/latest/array-creation.html#store-dask-arrays)):

```python
import dask.array as da
import zarr as zr
c = (2, 2)
d = da.ones((10, 11), chunks=c)
z1 = zr.open_array('lazy.zarr', shape=d.shape, dtype=d.dtype, chunks=c)
d1 = d.store(z1, compute=False, return_stored=True)
# fails with KeyError
```

GenevieveBuckley commented:
The problem seems to be here:

https://github.com/dask/dask/blob/850472dcf887bf1f1fefedf78c75f588ae65d53e/dask/array/core.py#L1028-L1034

Commenting out line 1029 will cause the same kind of failure, when passing a `compute=True` keyword argument.

GenevieveBuckley commented:
tl:dr 

## What is happening

Most tests are failing because we're calling `dask.array.store()` with `compute=False` and `return_stored=True`. That means this line isn't executed
https://github.com/dask/dask/blob/850472dcf887bf1f1fefedf78c75f588ae65d53e/dask/array/core.py#L1029

So then the dictionary `self.dask.layers` doesn't contain the keys we are expecting. We think there should be a key called `'load-store-6e49a6e2-7d68-11eb-9953-782b46cc48ef'` but instead we only get `dict_keys([139682423878752])`.

Separately, `test_constructor_plugin` fails because of this line in the test:

https://github.com/dask/dask/blob/850472dcf887bf1f1fefedf78c75f588ae65d53e/dask/array/tests/test_array_core.py#L3907

which means that when we make a new dask array, `self.dask` is a numpy array and not a `HighLevelGraph` (so we get an `AttributeError`).

## What will we do about it?

We need to work out:
!. What behaviour do we want in these situations?
2. How will we know when we're in that situation? (It's probably easy to check if `self.dask` is not a `HighLevelGraph`, but maybe it's a bit harder when it *is* a `HighLevelGraph` but just not one that has a dictionary with the keys we were expecting)

GenevieveBuckley commented:
@rjzamora & @crusaderky - Martin suggested you might be interested in taking a look at this. It's blocked until we've made a decision about the desired behaviour moving forward. Here's the summary:; https://github.com/dask/dask/pull/7309#issuecomment-791205031

crusaderky commented:
There's no guarantee that name matches an actually existing layer. It *normally* does, but there are exceptions. And you have to be very careful going down the route of "fixing" these exceptions, because you must avoid at all costs collisions in layer names.
This is, by the way, also a problem in ``__dask_layers__()``.
I think that your PR should try fetching a layer matching the name and gracefully do nothing if not found.

Please don't set attributes in a Python class that are not explicitly declared in the class itself. It's bad practice and a guarantee for things to fall apart when either (1) the class declares ``__slots__`` or (2) it is compiled with Cython or equivalent. I think that a generic dict container, e.g. Layer.info or symilar, would be a reasonable design.

> self.dask is a numpy array 

This makes no sense to me; how can a function that replaces a da.Array with a np.ndarray alter the da.Array.dask?

crusaderky commented on dask/array/core.py:
```diff
@@ -1156,6 +1156,10 @@ def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
 
+        self.dask.layers[name].type = type(self)
+        self.dask.layers[name].chunks = chunks
+        self.dask.layers[name].dtype = dtype
```
definitely meta type would be very useful

GenevieveBuckley commented:
Thanks for your comments @crusaderky 

1. I've added an `info` dictionary as an attribute to the `Layer` class with `self.info = {}`.
2. I've added an `if name in self.dask.layers` . (In retrospect this seems super obvious, I think at the time I was hesitant to add it just in case it was a sign I needed to "fix" things properly. Glad to hear that's not the case)
3. The reason we get a numpy array in `test_constructor_plugin` is that the array plugin passed is a function to call `.compute()` - the result of which is a numpy array. I'll explain this more clearly in another comment.

GenevieveBuckley commented:
Here is a more minimal example showing the reason for the failure in `test_constructor_plugin`.

```python
def test_genevieve():
    with dask.config.set(array_plugins=[lambda x: x.compute()]):
        x = da.ones(10, chunks=5)
        y = x + 1

    assert isinstance(y, np.ndarray)
```

Because the array plugin is a function calling `.compute()`, we get a numpy array which leads to unexpected results. Click to expand below to see the full details from pytest:
<details>
<summary>Details:</summary>

```
=================================== FAILURES ===================================
________________________________ test_genevieve ________________________________

    def test_genevieve():
        with dask.config.set(array_plugins=[lambda x: x.compute()]):
>           x = da.ones(10, chunks=5)

dask/array/tests/test_array_core.py:3895: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
dask/array/wrap.py:78: in wrap_func_shape_as_first_arg
    return Array(dsk, name, chunks, dtype=dtype, meta=kwargs.get("meta", None))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

cls = <class 'dask.array.core.Array'>
dask = <dask.highlevelgraph.HighLevelGraph object at 0x7fe43c9f7fa0>
name = 'ones-5dd10755422ede42ff13cddda545a499', chunks = ((5, 5),)
dtype = dtype('float64'), meta = array(2.5e-323), shape = None

    def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
        self = super(Array, cls).__new__(cls)
        assert isinstance(dask, Mapping)
        if not isinstance(dask, HighLevelGraph):
            dask = HighLevelGraph.from_collections(name, dask, dependencies=())
        self.dask = dask
        self.name = str(name)
        meta = meta_from_array(meta, dtype=dtype)
    
        if (
            isinstance(chunks, str)
            or isinstance(chunks, tuple)
            and chunks
            and any(isinstance(c, str) for c in chunks)
        ):
            dt = meta.dtype
        else:
            dt = None
        self._chunks = normalize_chunks(chunks, shape, dtype=dt)
        if self.chunks is None:
            raise ValueError(CHUNKS_NONE_ERROR_MESSAGE)
    
        self._meta = meta_from_array(meta, ndim=self.ndim, dtype=dtype)
    
        for plugin in config.get("array_plugins", ()):
            result = plugin(self)
            if result is not None:
                self = result
    
>       if name in self.dask.layers:
E       AttributeError: 'numpy.ndarray' object has no attribute 'dask'

dask/array/core.py:1159: AttributeError

```

</details>

crusaderky commented on dask/array/core.py:
```diff
@@ -1156,6 +1156,12 @@ def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
             if result is not None:
                 self = result
 
+        if hasattr(self, "dask") and name in self.dask.layers:
+            self.dask.layers[name].info["type"] = type(self)
+            self.dask.layers[name].info["chunk_type"] = type(meta)
+            self.dask.layers[name].info["chunks"] = chunks
+            self.dask.layers[name].info["dtype"] = dtype
+
```
instead of this
```suggestion
        try:
            layer = self.dask.layers[name]
        except (AttributeError, KeyError):
            # self is no longer an Array after applying the plugins, OR
            # self has been replaced by a collection where dask is not a HighLevelGraph, OR
            # name is not the top layer's name (this can happen after the layer is manipulated,
            # to avoid a collision)
            pass
        else:
            layer.info.update({"type": type(self), "chunk_type": type(meta), "chunks": chunks, "dtype": dtype})

```

crusaderky commented on dask/dataframe/core.py:
```diff
@@ -3609,6 +3609,13 @@ class DataFrame(_Frame):
+            self.dask.layers[name].info["type"] = type(self)
+            self.dask.layers[name].info["divisions"] = divisions
+            self.dask.layers[name].info["chunk_type"] = type(meta)
+
```
1. could use the same refactoring as da.Array
2. why nothing for Series?
3. dd does not have anything equivalent to array_plugins

GenevieveBuckley replied:
1. could use the same refactoring as da.Array - ok, done
2. why nothing for Series? - Good point, I've added `"series_dtypes": {col: meta[col].dtype for col in meta.columns}`
3. dd does not have anything equivalent to array_plugins - This looks like a comment and not a request. Thanks for mentioning it!

crusaderky commented on dask/highlevelgraph.py:
```diff
@@ -59,6 +59,7 @@ class Layer(collections.abc.Mapping):
     annotations: Optional[Mapping[str, Any]]
 
     def __init__(self, annotations: Mapping[str, Any] = None):
+        self.info = {}
```
1. please add type annotations above
2. please add documentation
3. you'll lose everything on clone(), cull(), and possibly some other methods

GenevieveBuckley replied:
1. please add type annotations above - added
2. please add documentation - added
3. you'll lose everything on clone(), cull(), and possibly some other methods - I think this is a comment, not a request to fix just now? If I've misunderstood please let me know.

crusaderky reviewed (changes requested):
Missing unit tests. Also, have there been measures on real-life use cases for storing the chunks information for each intermediate step this way? It can get pretty large pretty fast.

I'd like to hear from others what they think about the design - particularly since this new ``info`` dict is heavily overlapping with the ``annotations`` dict (but, unlike annotations, it will get lost in transit when moving to the distributed scheduler.

GenevieveBuckley commented:
> I'd like to hear from others what they think about the design - particularly since this new `info` dict is heavily overlapping with the `annotations` dict (but, unlike annotations, it will get lost in transit when moving to the distributed scheduler).

Are you suggesting putting this information into the `annotations` dict instead?

crusaderky commented:
@GenevieveBuckley That is a possibility but the potential performance implications of transferring the extra data to the scheduler are non-trivial. I'm suggesting that those who are heavily involved in the design of graph annotations should be involved in this design too

GenevieveBuckley commented:
> @GenevieveBuckley That is a possibility but the potential performance implications of transferring the extra data to the scheduler are non-trivial. I'm suggesting that those who are heavily involved in the design of graph annotations should be involved in this design too

@sjperkins do you have thoughts on this?

GenevieveBuckley commented:
cc @JSKenyon - I saw your name mentioned in the dask-blog issue on annotations, so perhaps you have thoughts here too

sjperkins commented:
Thanks for pinging me @GenevieveBuckley! This kind of metadata is super useful to include. @JSKenyon  and I have been experimenting with similar ideas.

I was considering the following structure:

```python
# setting collections=True mandates that collection metadata should be added to the layer
with dask.annotate(collections=True):
    A = da.ones((100, 100))
```

I like the idea of nesting the collection metadata within a "collection" (or "colleciton_meta") key as this identifies the task as being part of a collection. This provides the flavour of the idea:

```python
if dask.config.get("annotations", {}).get("collections", False):
    collection.layers[collection.name].annotations = {
         "collection": {
             "type": "Array",  # or "Dataframe"
             "chunks": collection.chunks,
             "chunk_type": collection.meta,
             # "dims": ("x", "y"),  # Could add xarray metadata too
             "dtype": collection.dtype,
          }
}
```

In particular we've found it useful to encode xarray dims too: There's possibility  for xarray integration  but this probably introduces too much scope creep for this PR. Pinging @dcherian.

What do you think?

GenevieveBuckley commented:
Can you comment on how this might impact performance @sjperkins? That's one concern @crusaderky raised

sjperkins commented:
> That is a possibility but the potential performance implications of transferring the extra data to the scheduler are non-trivial. 

With regard to concern regarding transmission, I think these annotations will be subject to the same performance characteristics of any annotation.

1. If the layer is materialised, an annotation will be [transmitted for every key](https://github.com/dask/dask/blob/e00f2fa2a8cb4c971dafc1f7b4c45caf53165879/dask/highlevelgraph.py#L206-L210), otherwise
2. If the layer is unmaterialized, only the annotation is transmitted and [expanded](https://github.com/dask/dask/blob/e00f2fa2a8cb4c971dafc1f7b4c45caf53165879/dask/highlevelgraph.py#L216) over keys on the scheduler.

I think the crux of the matter is whether (1) or (2) will be the general case going forward and this will be governed by whether graph optimisation occurs client (1) or server (2) side. I believe that (2) is the strategy going forward. @madsbk  please correct if the previous statement is wrong.

Note In the case of (1), I think that, as long as the annotation object is referenced multiple times in the annotation expansion, serialisation code should be smart enough to encode it once and insert references to the encoded version in the serialised stream. At least this seems to the case with pickle, msgpack seems to really struggle here. @madsbk, any thoughts or corrections on the msgpack use?

```python
In [43]: annotation = { 
    ...:     "collection": { 
    ...:         "type": "dask.array.core.Array", 
    ...:         "chunks": ((1000,)*10, (100,)*30, (4,)*40), 
    ...:         "dtype": "float64", 
    ...:     } 
    ...: }                                                                      

In [44]: len(pickle.dumps({k: annotation for k in range(100)}))                 
Out[44]: 695

In [45]: len(msgpack.dumps({k: annotation for k in range(100)}))                
Out[45]: 17003

In [46]: len(pickle.dumps({k: annotation for k in range(10000)}))               
Out[46]: 50059

In [47]: len(msgpack.dumps({k: annotation for k in range(10000)}))              
Out[47]: 1719619
```

In the case of (2), the serialised annotations are much smaller:

```python
In [48]: len(pickle.dumps(annotation))                                          
Out[48]: 292

In [49]: len(msgpack.dumps(annotation))                                         
Out[49]: 169
```

crusaderky commented:
> At least this seems to the case with pickle, msgpack seems to really struggle here.

msgpack has no concept of references to already-seen nodes so, short of implementing a custom extension, everything will be duplicated upon encoding.

mrocklin commented:
I think that we would not want these annotations to be expanded to every key at either stage.  The `chunks` in particular is likely to be large.  For large arrays this would be an n-squared cost.

mrocklin commented:
(thank you for raising this concern by the way).  

It would be nice to reuse the annotations machinery for this, but the "copy annotations to every task" policy seems to make that hard.

sjperkins commented:
@mrocklin

> I think that we would not want these annotations to be expanded to every key at either stage. The chunks in particular is likely to be large. For large arrays this would be an n-squared cost.

> (thank you for raising this concern by the way).
> It would be nice to reuse the annotations machinery for this, but the "copy annotations to every task" policy seems to make that hard.

I disagree, in principle a single layer layer annotation object should be re-usable over all keys as in the example below:

```python
In [16]: annotation = {  
    ...:     ...:     "collection": {  
    ...:     ...:         "type": "dask.array.core.Array",  
    ...:     ...:         "chunks": ((1000,)*10, (100,)*30, (4,)*40),  
    ...:     ...:         "dtype": "float64",  
    ...:     ...:     }  
    ...:     ...: }                                                             

In [17]: A = {('a', i): annotation for i in range(10000)}                       
In [18]: assert len(set(id(v["collection"]) for v in A.values())) == 1                        
In [19]: assert len(set(id(v["collection"]["chunks"]) for v in A.values())) == 1
```

So expanding this particular on the scheduler shouldn't be any more onerous in terms of memory than the existing scheduler annotation expansion system.  Additionally, if the following system is used, it becomes an opt-in for the user:

```python
# setting collections=True mandates that collection metadata should be added to the layer
with dask.annotate(collections=True):
    A = da.ones((100, 100))
```

sjperkins commented:
My concern is mainly for the case where the layer is materialised client-side before transmission ot the scheduler.

mrocklin commented:
Even if we don't do it today, we'll eventually want to send annotations to the workers.  Even though Python may save us on the scheduler side today, it won't save us as we try to communicate tasks to the workers.

I'm not saying that we can't get everything that we want here.  I'm just saying that there are some constraints that we'll have to think about.  For example, maybe we don't expand all annotations to all tasks.  Maybe there are layer-level annotations and task-level annotations.  There are plenty of solutions here.  We are in a space where we need to be mindful/careful though.

sjperkins commented:
> Even if we don't do it today, we'll eventually want to send annotations to the workers. Even though Python may save us on the scheduler side today, it won't save us as we try to communicate tasks to the workers.

I'm now imagining the scenario  where Layers are transmitted wholesale to a worker? My initial reaction is that this is too coarse-grained, but I haven't deeply considered the implications. Expanding on the existing strategy, I was imagining that this kind of dictionary would be transmitted by the scheduler to the worker for each task:

```python
{
     "func": dumps_function(function),
     "args": dumps(args),
     "kwargs": dumps(kwargs),
     "annotations": dumps(annotations),
}

```

There'd be an extra annotation field for each task, but if similar tasks with the same function and annotations are batched for worker transmission (as they will be for Layers) a sensible serialisation protocol should serialise the first occurrence and then refer to that serialised version for subsequent occurrences.

I'm fairly concerned that msgpack serialises duplicates occurrences individually, instead of inserting references in the serialisation stream as performed by pickle:

```python
In [49]: len(pickle.dumps(['deadbeef']*1000))                                                                                                                                                               
Out[49]: 2021
In [50]: len(msgpack.dumps(['deadbeef']*1000))                                                                                                                                                              
Out[50]: 9003
```

Have these implications been considered now that msgpack is used more extensively in [distributed](https://github.com/dask/distributed/pull/4531/files)?


> I'm not saying that we can't get everything that we want here. I'm just saying that there are some constraints that we'll have to think about. For example, maybe we don't expand all annotations to all tasks. 

Note that the current behaviour is to expand annotations over all keys on the scheduler: https://github.com/dask/dask/blob/e00f2fa2a8cb4c971dafc1f7b4c45caf53165879/dask/highlevelgraph.py#L216:

The invariants in the example still hold:

```python
In [27]: annotation = {  
    ...:     ...:     "collection": {  
    ...:     ...:         "type": "dask.array.core.Array",  
    ...:     ...:         "chunks": ((1000,)*10, (100,)*30, (4,)*40),  
    ...:     ...:         "dtype": "float64",  
    ...:     ...:     }  
    ...:     ...: }                 
In [28]: A = dict.fromkeys((('a', i) for i in range(1000)), annotation)         
In [29]: assert len(set(id(v["collection"]) for v in A.values())) == 1          
In [30]: assert len(set(id(v["collection"]["chunks"]) for v in A.values())) == 1
```

> Maybe there are layer-level annotations and task-level annotations. There are plenty of solutions here. We are in a space where we need to be mindful/careful though.

I hadn't considered a return to task-level annotations given the challenges with the initial attempts -- maybe there'll be an easier way forward in the future.

This doesn't prevent us changing behaviour on the scheduler in future, but if current behaviour expands annotations over collection keys on the scheduler, why not re-use the annotation mechanism to transmit collection information?  Otherwise, a separate transmission mechanism will be required to transmit layer information. In my experience the above collection annotation approach is both simple and effective.

sjperkins commented:
>  a sensible serialisation protocol should serialise the first occurrence and then refer to that serialised version for subsequent occurrences.

I'm saying based on my reading of protocol buffers and capnproto and my experience of pickle.

mrocklin commented:
> I'm now imagining the scenario where Layers are transmitted wholesale to a worker? 

No, I can't think of a case where that would occur.  But there may be cases where we want to annotate the layers without annotating the underlying tasks.  Showing the `chunks` of an array layer is such a case  I think.

GenevieveBuckley commented:
> I think that we would not want these annotations to be expanded to every key at either stage. The `chunks` in particular is likely to be large. For large arrays this would be an n-squared cost.

> But there may be cases where we want to annotate the layers without annotating the underlying tasks. Showing the `chunks` of an array layer is such a case I think.

In a lot of cases, I don't think users actually want this level of information. At least, not all of the time. With arrays, it's common to have uniformly sized chunks, perhaps with some funny sized ones towards the edges. Can we separate this somehow into:
1. What is the typical chunk size at this point in the computation? This is probably what people want to know first, then...
2. What is the size of *all* of the chunks?

I don't know if it's easy to create an answer to (1), but I think that's the first question people try and answer when troubleshooting, even if you start out by giving them all the detailed chunk sizes.

mrocklin commented:
This conversation has stalled, mostly due to future-leaning concerns about how this information that we're adding to the layers on the client side could affect downstream submission on the scheduler and the workers.  However, these concerns aren't yet valid today (we don't currently care about this information on the scheduler or workers yet, and so it would be good if these concerns did not block us).  

Short term, I recommend that we procede without using the current annotations machinery, and instead put this information somewhere else, something like `collection_annotations`.  We can just ignore these annotations completely when shipping things off to the scheduler for now.  If in the future we do want to send this data (which sounds like fun to me) then we can address the issues about overly replicating this data then.

@sjperkins are you ok with deferring the broader question here and letting this work go on?  If so, are you ok with putting this metadata somewhere other than annotations for the short-term?

sjperkins commented:
Apologies I did not realise this issue was blocked. I'm happy for this to proceed and defer the broader question.

GenevieveBuckley commented:
Ok I've tried out the `collection_annotations` approach suggested. Can I ask you all to take a look & see what you think now?

(If any of the CI tests fail, I'll come back and address those - it's late on Friday here so I should stop working now)

mrocklin commented:
> Ok I've tried out the collection_annotations approach suggested. Can I ask you all to take a look & see what you think now?

> (If any of the CI tests fail, I'll come back and address those - it's late on Friday here so I should stop working now)

@sjperkins would you be interested in taking another look through this?

sjperkins commented on dask/highlevelgraph.py:
```diff
@@ -57,12 +57,40 @@ class Layer(collections.abc.Mapping):
+        else:
+            self.collection_annotations = copy.copy(
+                config.get("collection_annotations", {})
+            )
```
Is the intent here to support the following?

```python
wiith dask.config.set(collection_annotations={"blah": 1}:
   ...
```

I wonder if this should be documented, although I might say not if the intention is for ingestion by internal visualization routines.

mrocklin replied:
If we document it then we need to support it.  Right now all we need is for this to be supported by internal functions.  I don't see a reason to intentionally expose or advertise this out to users.  Given that this part of the codebase is under churn I'm inclined to not promise much to users here.

GenevieveBuckley replied:
> intention is for ingestion by internal visualization routines

I'd say this description is accurate.

sjperkins reviewed (commented):
My understand of this PR is that it exists to provide information to dask visualization routines (`dask.visualize`?). I think the approach for the Array and Dataframe collections is fine.

What about Delayed and Bag? I believe Delayed objects have a HLG with layers but Bags do not. I'd be happy with attaching annotations to Delayed objects only, perhaps just with a `type` attribute. Are there any other attributes appropriate for Delayed?

mrocklin commented:
> I think the approach for the Array and Dataframe collections is fine.

Woo

> What about Delayed and Bag? I believe Delayed objects have a HLG with layers but Bags do not. I'd be happy with attaching annotations to Delayed objects only, perhaps just with a type attribute. Are there any other attributes appropriate for Delayed?

Bag could use high level graphs if anyone spent the time to do it.  I think that we're establishing a protocol here that can be extended in the future.

GenevieveBuckley commented:
> My understand of this PR is that it exists to provide information to dask visualization routines 
Yes, this is accurate.

> I think the approach for the Array and Dataframe collections is fine.
> What about Delayed and Bag?

Since this PR supports experimental mucking around with visualizations, I think it makes sense to defer any decisions for Delayed & Bag until after we've had a chance to try it and see if it's a useful thing we want to do.

GenevieveBuckley commented:
Thank you for looking over this again @sjperkins (and thank you @crusaderky for your earlier review, too)

jrbourbeau commented on dask/array/core.py:
```diff
@@ -1156,6 +1156,24 @@ def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
+            layer.collection_annotations.update(
+                {
+                    "type": type(self),
+                    "chunk_type": type(meta),
```
Nitpick: What you have is totally fine and `type(meta)` should be the same as `type(self._meta)`. Though this may change in the future if we add extra code between where `meta` and `self._meta` are defined above. Unless there's a reason not to (which I may be missing) then I'd prefer to use `self._meta` here to keep things more robust to future changes. 

```suggestion
                    "chunk_type": type(self._meta),
```

jrbourbeau commented on dask/highlevelgraph.py:
```diff
@@ -57,12 +57,40 @@ class Layer(collections.abc.Mapping):
+    def __init__(
+        self,
+        annotations: Mapping[str, Any] = None,
+        collection_annotations: Mapping[str, Any] = {},
```
It would be good to avoid a mutable default for a keyword argument. What you have here is totally fine today, but can lead to unexpected behavior in the future which is hard to track down. 

```suggestion
        collection_annotations: Mapping[str, Any] = None,
```

jrbourbeau commented on dask/array/core.py:
```diff
@@ -1156,6 +1156,24 @@ def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
+                {
+                    "type": type(self),
+                    "chunk_type": type(meta),
+                    "chunks": chunks,
+                    "dtype": dtype,
```
Similar comment here for using `self.chunks` and `self.dtype`. I'll defer to you about what you think is best to keep around for visualization purposes. Though note that while `chunks` can be things like `"auto"` and `2`, `self.chunks` will be the expanded chunks for the full array like `((2, 2, 2, 2, 2),)`.

GenevieveBuckley replied:
There are several cases where using `self.dtype` like you suggest produces failing tests in `dask/test/test_gufunc.py`

These tests fail because in these cases `meta` / `self._meta` is a tuple. Tuples do not have a `.dtype` attribute, so when we look up `self.dtype` we get an error. 

```python
    @property
    def dtype(self):
>       return self._meta.dtype
```

```
dask/array/tests/test_gufunc.py::test_apply_gufunc_01 FAILED             [ 12%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_01b FAILED            [ 14%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_output_dtypes_string_many_outputs[False] FAILED [ 20%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_output_dtypes_string_many_outputs[True] FAILED [ 22%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_elemwise_02 FAILED    [ 34%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_elemwise_loop FAILED  [ 38%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_elemwise_core FAILED  [ 40%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_two_scalar_output FAILED [ 42%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_two_mixed_outputs FAILED [ 44%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_broadcasting_loopdims FAILED [ 58%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_infer_dtype FAILED    [ 66%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_axis_02 FAILED        [ 72%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_axis_02b FAILED       [ 74%]
dask/array/tests/test_gufunc.py::test_preserve_meta_type FAILED          [ 96%]
dask/array/tests/test_gufunc.py::test_apply_gufunc_with_meta FAILED      [ 98%]
dask/array/tests/test_gufunc.py::test_as_gufunc_with_meta FAILED         [100%]
```

There are no problems with `self.chunks`, and in fact it's preferable to have the full expanded chunks listed instead of `"auto"`.

jrbourbeau replied:
Thanks for looking into this @GenevieveBuckley. Let's use `dtype` then -- we can always make updates as needed later on

jrbourbeau reviewed (commented):
Thanks for all your work on this @GenevieveBuckley! I've left a few small final comments, but overall this looks ready to merge

jrbourbeau reviewed (approved):
Thanks @GenevieveBuckley for your work on this (and @sjperkins @crusaderky for reviewing)! I'm looking forward to seeing how we can use this new information

GenevieveBuckley commented:
@freyam this PR might be relevant for your GSOC project later on.

You don't have to read everything or understand all of the discussion (obviously things changed as we went, so the first parts might not make much sense.

The part that's relevant for you is that we now have a new dictionary called `collection_annotations` (alongside the [existing `annotations`](https://docs.dask.org/en/latest/api.html?highlight=annotations#dask.annotate) dictionary). The new `collection_annotations` dictionary is an attribute of the dask layer, and holds information about the characteristics of dask arrays or dataframes (the info you get is slightly different depending on whether it's an array or a dataframe). We might try to integrate this information into our visualizations.

```python
In [1]: import dask.array as da

In [2]: arr = da.random.random((100,100), chunks=(10,10))

In [3]: arr.dask.layers
Out[3]: {'random_sample-8c39afc91c532043b96497a2f2fa9875': <dask.highlevelgraph.MaterializedLayer at 0x7fe4b4339640>}

In [4]: mylayer = arr.dask.layers['random_sample-8c39afc91c532043b96497a2f2fa987
   ...: 5']

In [5]: mylayer
Out[5]: <dask.highlevelgraph.MaterializedLayer at 0x7fe4b4339640>

In [6]: mylayer.collection_annotations
Out[6]: 
{'type': dask.array.core.Array,
 'chunk_type': numpy.ndarray,
 'chunks': ((10, 10, 10, 10, 10, 10, 10, 10, 10, 10),
  (10, 10, 10, 10, 10, 10, 10, 10, 10, 10)),
 'dtype': None}

```

rjzamora commented:
>Thanks @GenevieveBuckley for your work on this (and @sjperkins @crusaderky for reviewing)! I'm looking forward to seeing how we can use this new information

Just a note:  I am starting to (slowly) work through a design doc for general column-projection in Dask-Dataframe, and I am getting the sense that `collection_annotations` may be the right place to store `Layer`-wise column-dependency properties. That is, I am thinking that [optimize_dataframe_getitem](https://github.com/dask/dask/blob/cae3f8ffcb0891d8b04e1900c34ba4433c55945f/dask/dataframe/optimize.py#L49) can be extended to work accross multiple layers if common Dataframe operations are modified to (optionally) store the required input and output columns for the generated Layers.  If all Layers in a HLG include this metadata, and the root Layer is a DataFrameIOLayer, then column projection becomes relatively simple.

mrocklin commented:
Rick, I think that it is also reasonable for you to make a DataframeLayer
class that defines a consistent set of attributes.  I don't think that you
need to restrict yourself to this dictionary and can probably use normal
Python attributes.  However, I do think that DataframeLayer should probably
populate this dict secondarily if someone asks for it.

On Thu, Jul 1, 2021 at 6:22 PM Richard (Rick) Zamora <
***@***.***> wrote:

> Thanks @GenevieveBuckley <https://github.com/GenevieveBuckley> for your
> work on this (and @sjperkins <https://github.com/sjperkins> @crusaderky
> <https://github.com/crusaderky> for reviewing)! I'm looking forward to
> seeing how we can use this new information
>
> Just a note: I am starting to (slowly) work through a design doc for
> general column-projection in Dask-Dataframe, and I am getting the sense
> that collection_annotations may be the right place to store Layer-wise
> column-dependency properties. That is, I am thinking that
> optimize_dataframe_getitem
> <https://github.com/dask/dask/blob/cae3f8ffcb0891d8b04e1900c34ba4433c55945f/dask/dataframe/optimize.py#L49>
> can be extended to work accross multiple layers if common Dataframe
> operations are modified to (optionally) store the required input and output
> columns for the generated Layers. If all Layers in a HLG include this
> metadata, and the root Layer is a DataFrameIOLayer, then column projection
> becomes relatively simple.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/7309#issuecomment-872608525>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACKZTBIR634735EFHKJYFTTVT2E5ANCNFSM4YQPLTNA>
> .
>

rjzamora commented:
>Rick, I think that it is also reasonable for you to make a DataframeLayer
class that defines a consistent set of attributes.

Right - That is certainly plan "A", but I am starting to *consider* alternatives as I struggle to decide on a "clean" sub-classing design :)

mrocklin commented:
Maybe share your thoughts in an issue?  Perhaps we can think through a good
design together.

On Thu, Jul 1, 2021, 5:34 PM Richard (Rick) Zamora ***@***.***>
wrote:

> Rick, I think that it is also reasonable for you to make a DataframeLayer
> class that defines a consistent set of attributes.
>
> Right - That is certainly plan "A", but I am starting to *consider*
> alternatives as I struggle to decide on a "clean" sub-classing design :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/7309#issuecomment-872631978>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACKZTCDCMMX5QC2EBG6M4TTVUCR3ANCNFSM4YQPLTNA>
> .
>
